### ISSUE TITLE
perf issues in gulp-typescript

### ISSUE THREAD

**Issue by vladima** [ivogabe/gulp-typescript#495]
Initially reported [here](https://github.com/Microsoft/web-build-tools/issues/128).
Repro steps:
```
git clone https://github.com/OfficeDev/office-ui-fabric-react.git
cd office-ui-fabric-react/
git checkout 8332486ec9801b78bf398f4d85164c8fc89b79c3
npm install
npm start
```
once server is started stop it
```
cd packages/office-ui-fabric-react
```
Running `tsc` directly takes around 5.5 seconds on my machine
```
vladima@vladima-HP:~/Sources/git/office-ui-fabric-react/packages/office-ui-fabric-react$ ./node_modules/.bin/tsc --diagnostics
Files:           381
Lines:         49882
Nodes:        214682
Identifiers:   66158
Symbols:       84045
Types:         23312
Memory used: 185966K
I/O read:      0.02s
I/O write:     0.07s
Parse time:    0.90s
Bind time:     0.50s
Check time:    2.48s
Emit time:     1.39s
Total time:    5.28s
```
However `gulp ts` only finishes in 1.17 min
```
vladima@vladima-HP:~/Sources/git/office-ui-fabric-react/packages/office-ui-fabric-react$ ./node_modules/.bin/gulp ts
[11:30:42] Using gulpfile ~/Sources/git/office-ui-fabric-react/packages/office-ui-fabric-react/gulpfile.js
[11:30:42] Starting gulp
[11:30:42] Starting 'ts'...
[11:30:42] Starting subtask 'typescript'...
[11:30:42] [typescript] TypeScript version: 2.1.6
[11:31:52] Finished subtask 'typescript' after 1.15 min
[11:31:52] Finished 'ts' after 1.15 min
[11:31:52] ==================[ Finished ]==================
[11:31:53] Project office-ui-fabric-react version: 1.14.1
[11:31:53] Build tools version: 2.4.0
[11:31:53] Node version: v7.7.1
[11:31:53] Total duration: 1.17 min
```

At glance it seems that issue originates from [these](https://github.com/ivogabe/gulp-typescript/blob/master/lib/compiler.ts#L64-L92) bits and partially caused by the fact that `Program` does not cache semantic and declaration diagnostics and declaration diagnostics are pretty expensive to compute since to do it compiler has to effectively generate all declaration files and drop them on the floor.

- `getPreEmitDiagnostics` [will get](https://github.com/ivogabe/gulp-typescript/blob/master/lib/compiler.ts#L64) combined syntactic/semantic and declaration diagnostics and return them as one list
- then code will [query](https://github.com/ivogabe/gulp-typescript/blob/master/lib/compiler.ts#L67-L70) for syntactic/semantic/options/global diagnostics just to get them separately
- then declaration diagnostics will be [requested one more time](https://github.com/ivogabe/gulp-typescript/blob/master/lib/compiler.ts#L67-L73)
- then in [program.emit](https://github.com/ivogabe/gulp-typescript/blob/master/lib/compiler.ts#L77) all diagnostics will be requested one more time if `noEmitOnError` is set.
- and as a last step it will [emit one file at a time](https://github.com/ivogabe/gulp-typescript/blob/master/lib/compiler.ts#L85-L90) and if `noEmitOnError` is set then every emit will lead to getting declaration errors (effectively regenerate declaration files for all files in program).

**Comment by vladima** [comment]
https://github.com/Microsoft/TypeScript/pull/14516 should address issues when diagnostics requested multiple times. However there are few things that can be handled in gulp-typescript:
- do not emit files multiple times
- minimize number of calls to get diagnostics

**Comment by ivogabe** [comment]
Thanks for the detailed info. Can you take a look at 7c91b82c6ba1c268c3f19af88661b84b2d5e56b9? I'll check whether that fixes the performance issues in the referenced repo.

**Comment by ivogabe** [comment]
I measured it, time dropped from 43 seconds to 40 seconds so that's not significant. Are more changes required, or will Microsoft/TypeScript#14516 fix the rest of this issue?

**Comment by vladima** [comment]
I'm curious if this snippet:
```ts
			for (const fileName of this.host.input.getFileNames(true)) {
				const file = this.project.input.getFile(fileName);

				this.emitFile(result, currentDirectory, file);
			}
```
can be replaced with single call to `program.emitFile`?

**Comment by ivogabe** [comment]
We need to have a relation between the input and the output files, and this way we can get that very easily. In 2.x some code was used that guessed the input file name given an output file name, but that was very buggy. All different settings (`outDir`, `out` etc but also location of `tsconfig`, base directory ..) make it very hard to do that correctly.

**Comment by vladima** [comment]
the [signature](https://github.com/Microsoft/TypeScript/blob/master/src/compiler/types.ts#L2304) of `WriteFileCallback` is:
```ts
  export interface WriteFileCallback {
        (fileName: string, data: string, writeByteOrderMark: boolean, onError?: (message: string) => void, sourceFiles?: SourceFile[]): void;
    }
```

so it accepts array of original source files as the last parameter.

**Comment by ivogabe** [comment]
Will take a look at that tomorrow. I guess that the array contains all files if `outFile` or `out` is set, and exactly one otherwise?

**Comment by vladima** [comment]
yes, that is correct. I think this last parameter was added in TS 2.0

**Comment by ivogabe** [comment]
With 489f54ca14328c31c761b82508982a8b9083b199 it now compiles in 7.5 seconds! The tests here in gulp-typescript also run faster now. I'll do some more testing tomorrow.

**Comment by dzearing** [comment]
@ivogabe Thanks a ton for looking at this. Totally looking forward for this to drop, we'll update our build stuff fast and get this in, save a lot of people a lot of time.

**Comment by lalarsson** [comment]
@ivogabe Any idea of when this fix might be released? Would be awesome to get this into our build environment :)

**Comment by ivogabe** [comment]
I planned to release it with some other new features, but those take longer. I'll put those features in a later release and release this fix today.

**Comment by ivogabe** [comment]
Released as 3.1.6 🎉

**Comment by jcristovao** [comment]
Hi!
I'm afraid I'm still experiencing a 10x slowdown when using `gulp-typescript` vs running `tsc` directly, even using 3.1.6

First of all, my `package.json` does not have either a `outFile` entry or a `outDir`, since I'm afterwards passing the output through babel to convert the es6 modules (this is just for node), while when running `tsc` I'm using the `outDir` flag. Does that make a difference? From your discussion I kind of get that it might, but I'm not sure - as I'm also not sure what should I even put in `outDir`, since I'm using `gulp.dest` to set the final destination.

I'm also using `gulp-debug` to assert the time taken on each file, so I'm fairly sure the delay is on typescript compilation, not babel.

Relevant part:

```
    .pipe(sourcemaps.init())
    .pipe(tsFiles)
    .pipe(debug({title:'[TS    ]'}))
    .pipe(tsProject(ts.reporter.longReporter()))  // if ts files, transpile with typescript
    .pipe(tsFiles.restore)
    .pipe(debug({title:'[BABEL ]'}))
    .pipe(babel({ babelrc: false
                , presets: [ 'eslatest-node6'
                           , 'react' ]
                , plugins: [ 'transform-es2015-destructuring'
                           , 'transform-es2015-parameters'
                           , 'transform-object-rest-spread'
                           ]
                 })) 
```

Any clues?
